File: app/src/main/java/io/intrepid/contest/base/BasePresenter.java
package io.intrepid.contest.base;

import android.support.annotation.NonNull;

import io.intrepid.contest.rest.RestApi;
import io.intrepid.contest.settings.PersistentSettings;
import io.reactivex.ObservableTransformer;
import io.reactivex.Scheduler;
import io.reactivex.disposables.CompositeDisposable;

public abstract class BasePresenter<T extends BaseContract.View> implements BaseContract.Presenter<T> {

    protected final CompositeDisposable disposables = new CompositeDisposable();

    @NonNull
    protected final Scheduler ioScheduler;
    @NonNull
    protected final Scheduler uiScheduler;
    @NonNull
    protected final PersistentSettings persistentSettings;
    @NonNull
    protected final RestApi restApi;
    protected T view;
    private boolean isViewBound = false;

    public BasePresenter(@NonNull T view, @NonNull PresenterConfiguration configuration) {
        this.view = view;
        this.ioScheduler = configuration.getIoScheduler();
        this.uiScheduler = configuration.getUiScheduler();
        this.persistentSettings = configuration.getPersistentSettings();
        this.restApi = configuration.getRestApi();
    }

    @Override
    public void onViewCreated() {

    }

    @Override
    public final void bindView(@NonNull T view) {
        this.view = view;

        if (!isViewBound) {
            onViewBound();
            isViewBound = true;
        }
    }

    protected void onViewBound() {

    }

    @Override
    public final void unbindView() {
        disposables.clear();
        this.view = null;

        if (isViewBound) {
            onViewUnbound();
            isViewBound = false;
        }
    }

    protected void onViewUnbound() {

    }

    /**
     * Note: The view will be null at this point. This method is for any additional cleanup that's not handled
     * by the CompositeSubscription
     */
    @Override
    public void onViewDestroyed() {

    }

    protected <R> ObservableTransformer<R, R> subscribeOnIoObserveOnUi() {
        return observable -> observable.subscribeOn(ioScheduler).observeOn(uiScheduler);
    }


}

File: app/src/main/java/io/intrepid/contest/base/BasePresenter.kt
package io.intrepid.contest.base

import io.intrepid.contest.rest.RestApi
import io.intrepid.contest.settings.PersistentSettings
import io.reactivex.ObservableTransformer
import io.reactivex.Scheduler
import io.reactivex.disposables.CompositeDisposable

abstract class BasePresenter<T : BaseContract.View>(protected var view: T?, configuration: PresenterConfiguration) : BaseContract.Presenter<T> {

    protected val disposables = CompositeDisposable()

    protected val ioScheduler: Scheduler = configuration.ioScheduler
    protected val uiScheduler: Scheduler = configuration.uiScheduler
    protected val persistentSettings: PersistentSettings = configuration.persistentSettings
    protected val restApi: RestApi = configuration.restApi
    private var isViewBound = false

    override fun onViewCreated() {

    }

    override fun bindView(view: T) {
        this.view = view

        if (!isViewBound) {
            onViewBound()
            isViewBound = true
        }
    }

    protected open fun onViewBound() {

    }

    override fun unbindView() {
        disposables.clear()
        this.view = null

        if (isViewBound) {
            onViewUnbound()
            isViewBound = false
        }
    }

    protected fun onViewUnbound() {

    }

    /**
     * Note: The view will be null at this point. This method is for any additional cleanup that's not handled
     * by the CompositeSubscription
     */
    override fun onViewDestroyed() {

    }

    protected fun <R> subscribeOnIoObserveOnUi(): ObservableTransformer<R, R> {
        return ObservableTransformer { observable -> observable.subscribeOn(ioScheduler).observeOn(uiScheduler) }
    }

}
